File: mysql-async/src/test/scala/com/github/mauricio/async/db/mysql/PreparedStatementsSpec.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.mysql

import java.sql.Timestamp
import java.util.concurrent.TimeUnit
import org.joda.time._
import org.specs2.mutable.Specification
import scala.concurrent.duration.Duration

class PreparedStatementsSpec extends Specification with ConnectionHelper {

  "connection" should {

    "be able to execute prepared statements" in {

      withConnection {
        connection =>
          val result = executePreparedStatement(connection, "select 1 as id , 'joe' as name").rows.get

          result(0)("name") === "joe"
          result(0)("id") === 1
          result.length === 1

          val otherResult = executePreparedStatement(connection, "select 1 as id , 'joe' as name").rows.get

          otherResult(0)("name") === "joe"
          otherResult(0)("id") === 1
          otherResult.length === 1
      }

    }

    "be able to detect a null value in a prepared statement" in {

      withConnection {
        connection =>
          val result = executePreparedStatement(connection, "select 1 as id , 'joe' as name, NULL as null_value").rows.get

          result(0)("name") === "joe"
          result(0)("id") === 1
          result(0)("null_value") must beNull
          result.length === 1

      }

    }

    "be able to select numbers and process them" in {

      withConnection {
        connection =>
          executeQuery(connection, createTableNumericColumns)
          executeQuery(connection, insertTableNumericColumns)
          val result = executePreparedStatement(connection, "SELECT * FROM numbers").rows.get(0)

          result("number_tinyint").asInstanceOf[Byte] === -100
          result("number_smallint").asInstanceOf[Short] === 32766
          result("number_mediumint").asInstanceOf[Int] === 8388607
          result("number_int").asInstanceOf[Int] === 2147483647
          result("number_bigint").asInstanceOf[Long] === 9223372036854775807L
          result("number_decimal") === BigDecimal(450.764491)
          result("number_float") === 14.7F
          result("number_double") === 87650.9876
      }

    }

    "be able to select from a table with timestamps" in {

      withConnection {
        connection =>
          executeQuery(connection, createTableTimeColumns)
          executeQuery(connection, insertTableTimeColumns)
          val result = executePreparedStatement(connection, "SELECT * FROM posts").rows.get(0)

          val date = result("created_at_date").asInstanceOf[LocalDate]

          date.getYear === 2038
          date.getMonthOfYear === 1
          date.getDayOfMonth === 19

          val dateTime = result("created_at_datetime").asInstanceOf[LocalDateTime]
          dateTime.getYear === 2013
          dateTime.getMonthOfYear === 1
          dateTime.getDayOfMonth === 19
          dateTime.getHourOfDay === 3
          dateTime.getMinuteOfHour === 14
          dateTime.getSecondOfMinute === 7

          val timestamp = result("created_at_timestamp").asInstanceOf[LocalDateTime]
          timestamp.getYear === 2020
          timestamp.getMonthOfYear === 1
          timestamp.getDayOfMonth === 19
          timestamp.getHourOfDay === 3
          timestamp.getMinuteOfHour === 14
          timestamp.getSecondOfMinute === 7

          result("created_at_time") === Duration(3, TimeUnit.HOURS) + Duration(14, TimeUnit.MINUTES) + Duration(7, TimeUnit.SECONDS)

          val year = result("created_at_year").asInstanceOf[Short]

          year === 1999
      }

    }

    "it should be able to bind statement values to the prepared statement" in {

      withConnection {
        connection =>
          val insert =
            """
              |insert into numbers (
              |number_tinyint,
              |number_smallint,
              |number_mediumint,
              |number_int,
              |number_bigint,
              |number_decimal,
              |number_float,
              |number_double
              |) values
              |(
              |?,
              |?,
              |?,
              |?,
              |?,
              |?,
              |?,
              |?)
            """.stripMargin


          val byte: Byte = 10
          val short: Short = 679
          val mediumInt = 778
          val int = 875468
          val bigInt = BigInt(100007654)
          val bigDecimal = BigDecimal("198.657809")
          val double = 98.765
          val float = 432.8F

          executeQuery(connection, this.createTableNumericColumns)
          executePreparedStatement(connection,
            insert,
            byte,
            short,
            mediumInt,
            int,
            bigInt,
            bigDecimal,
            float,
            double)

          val row = executePreparedStatement(connection, "SELECT * FROM numbers").rows.get(0)

          row("number_tinyint") === byte
          row("number_smallint") === short
          row("number_mediumint") === mediumInt
          row("number_int") === int
          row("number_bigint") === bigInt
          row("number_decimal") === bigDecimal
          row("number_float") === float
          row("number_double") === double

      }

    }

    "bind parameters on a prepared statement" in {

      val create = """CREATE TEMPORARY TABLE posts (
                     |       id INT NOT NULL AUTO_INCREMENT,
                     |       some_text TEXT not null,
                     |       primary key (id) )""".stripMargin

      val insert = "insert into posts (some_text) values (?)"
      val select = "select * from posts"

      withConnection {
        connection =>
          executeQuery(connection, create)
          executePreparedStatement(connection, insert, "this is some text here")
          val row = executePreparedStatement(connection, select).rows.get(0)

          row("id") === 1
          row("some_text") === "this is some text here"

          val queryRow = executeQuery(connection, select).rows.get(0)

          queryRow("id") === 1
          queryRow("some_text") === "this is some text here"


      }
    }

    "bind timestamp parameters to a table" in {

      val insert =
        """
          |insert into posts (created_at_date, created_at_datetime, created_at_timestamp, created_at_time, created_at_year)
          |values ( ?, ?, ?, ?, ? )
        """.stripMargin

      val date = new LocalDate(2011, 9, 8)
      val dateTime = new LocalDateTime(2012, 5, 27, 15, 29, 55)
      val timestamp = new Timestamp(dateTime.toDateTime.getMillis)
      val time = Duration( 3, TimeUnit.HOURS ) + Duration( 5, TimeUnit.MINUTES ) + Duration(10, TimeUnit.SECONDS)
      val year = 2012

      withConnection {
        connection =>
          executeQuery(connection, this.createTableTimeColumns)
          executePreparedStatement(connection, insert, date, dateTime, timestamp, time, year)
          val rows = executePreparedStatement(connection, "select * from posts where created_at_year > ?", 2011).rows.get

          rows.length === 1
          val row = rows(0)

          row("created_at_date") === date
          row("created_at_timestamp") === new LocalDateTime( timestamp.getTime )
          row("created_at_time") === time
          row("created_at_year") === year
          row("created_at_datetime") === dateTime

      }
    }

    "read a timestamp with microseconds" in {



      val create =
        """CREATE TEMPORARY TABLE posts (
       id INT NOT NULL AUTO_INCREMENT,
       created_at_timestamp TIMESTAMP(3) not null,
       created_at_time TIME(3) not null,
       primary key (id)
      )"""

      val insert =
        """INSERT INTO posts ( created_at_timestamp, created_at_time )
          | VALUES ( '2013-01-19 03:14:07.019', '03:14:07.019' )""".stripMargin

      val time = Duration(3, TimeUnit.HOURS ) +
        Duration(14, TimeUnit.MINUTES) +
        Duration(7, TimeUnit.SECONDS) +
        Duration(19, TimeUnit.MILLISECONDS)

      val timestamp = new LocalDateTime(2013, 1, 19, 3, 14, 7, 19)
      val select = "SELECT * FROM posts"

      withConnection {
        connection =>

          if ( connection.version < MySQLConnection.MicrosecondsVersion ) {
            pending(s"this version of MySQL (${connection.version}) does not support microseconds")
          } else {
            executeQuery(connection, create)
            executeQuery(connection, insert)
            val rows = executePreparedStatement( connection, select).rows.get

            val row = rows(0)

            row("created_at_time") === time
            row("created_at_timestamp") === timestamp

            val otherRow = executeQuery( connection, select ).rows.get(0)

            otherRow("created_at_time") === time
            otherRow("created_at_timestamp") === timestamp
          }

      }

    }

    "support prepared statement with a big string" in {

      val bigString = {
        val builder = new StringBuilder()
        for (i <- 0 until 400)
          builder.append( "0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789")
        builder.toString()
      }

      withConnection {
        connection =>
          executeQuery(connection, "CREATE TEMPORARY TABLE BIGSTRING( id INT NOT NULL AUTO_INCREMENT, STRING LONGTEXT, primary key (id))")
          executePreparedStatement(connection, "INSERT INTO BIGSTRING (STRING) VALUES (?)", bigString)
          val row = executePreparedStatement(connection, "SELECT STRING, id FROM BIGSTRING").rows.get(0)
          row("id") === 1
          val result = row("STRING").asInstanceOf[String]
          result === bigString
      }
    }

    "support setting null to a column" in {
      withConnection {
        connection =>
          executeQuery(connection, "CREATE TEMPORARY TABLE timestamps ( id INT NOT NULL, moment TIMESTAMP NULL, primary key (id))")
          executePreparedStatement(connection, "INSERT INTO timestamps (moment, id) VALUES (?, ?)", null, 10)
          val row = executePreparedStatement(connection, "SELECT moment, id FROM timestamps").rows.get(0)
          row("id") === 10
          row("moment") === null
      }
    }

    "support setting None to a column" in {
      withConnection {
        connection =>
          executeQuery(connection, "CREATE TEMPORARY TABLE timestamps ( id INT NOT NULL, moment TIMESTAMP NULL, primary key (id))")
          executePreparedStatement(connection, "INSERT INTO timestamps (moment, id) VALUES (?, ?)", None, 10)
          val row = executePreparedStatement(connection, "SELECT moment, id FROM timestamps").rows.get(0)
          row("id") === 10
          row("moment") === null
      }
    }

    "support setting Some(value) to a column" in {
      withConnection {
        connection =>
          executeQuery(connection, "CREATE TEMPORARY TABLE timestamps ( id INT NOT NULL, moment TIMESTAMP NULL, primary key (id))")
          val moment = LocalDateTime.now().withMillisOfDay(0) // cut off millis to match timestamp
          executePreparedStatement(connection, "INSERT INTO timestamps (moment, id) VALUES (?, ?)", Some(moment), 10)
          val row = executePreparedStatement(connection, "SELECT moment, id FROM timestamps").rows.get(0)
          row("id") === 10
          row("moment") === moment
      }
    }

  }

}

File: mysql-async/src/test/java/com/github/jasync/sql/db/mysql/PreparedStatementsSpec.kt
package com.github.jasync.sql.db.mysql

import com.github.jasync.sql.db.interceptor.MdcQueryInterceptorSupplier
import com.github.jasync.sql.db.interceptor.QueryInterceptor
import com.github.jasync.sql.db.invoke
import com.github.jasync.sql.db.util.map
import java.math.BigDecimal
import java.sql.Timestamp
import java.time.Duration
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.function.Supplier
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.slf4j.MDC

class PreparedStatementsSpec : ConnectionHelper() {

    @Test
    fun `be able to execute prepared statements`() {

        withConnection { connection ->
            val result = assertNotNull(executePreparedStatement(connection, "select 1 as id , 'joe' as name").rows)
            assertEquals(1, result.size)
            assertEquals("joe", result[0]["name"])
            assertEquals(1L, result[0]["id"])
            val otherResult = assertNotNull(executePreparedStatement(connection, "select 1 as id , 'joe' as name").rows)

            assertEquals(1, otherResult.size)
            assertEquals("joe", otherResult[0]["name"])
            assertEquals(1L, otherResult[0]["id"])
        }
    }

    @Test
    fun `be able to detect a null value()a prepared statement`() {

        withConnection { connection ->
            val result = assertNotNull(
                executePreparedStatement(
                    connection,
                    "select 1 as id , 'joe' as name, NULL as null_value"
                ).rows
            )
            assertEquals(1, result.size)
            assertEquals("joe", result[0]["name"])
            assertEquals(1L, result[0]["id"])
            assertNull(result[0]["null_value"])
        }
    }

    @Test
    fun `be able to select numbers and process them`() {

        withConnection { connection ->
            executeQuery(connection, createTableNumericColumns)
            executeQuery(connection, insertTableNumericColumns)
            val result =
                assertNotNull(assertNotNull(executePreparedStatement(connection, "SELECT * FROM numbers").rows)[0])

            assertEquals(-100, result["number_tinyint"] as Byte)
            assertEquals(32766, result["number_smallint"] as Short)
            assertEquals(8388607, result["number_mediumint"] as Int)
            assertEquals(2147483647, result["number_int"] as Int)
            assertEquals(9223372036854775807, result["number_bigint"] as Long)
            assertEquals(BigDecimal(450.764491).toDouble(), (result["number_decimal"] as BigDecimal).toDouble())
            assertEquals(14.7F, result["number_float"])
            assertEquals(87650.9876, result["number_double"])
        }
    }

    @Test
    fun `be able to select from a table with timestamps`() {

        withConnection { connection ->
            executeQuery(connection, createTableTimeColumns)
            executeQuery(connection, insertTableTimeColumns)
            val result =
                assertNotNull(assertNotNull(executePreparedStatement(connection, "SELECT * FROM posts").rows)[0])
            val date = result["created_at_date"] as LocalDate

            assertEquals(2038, date.year)
            assertEquals(1, date.monthValue)
            assertEquals(19, date.dayOfMonth)

            val dateTime = result["created_at_datetime"] as LocalDateTime

            assertEquals(2013, dateTime.year)
            assertEquals(1, dateTime.monthValue)
            assertEquals(19, dateTime.dayOfMonth)
            assertEquals(3, dateTime.hour)
            assertEquals(14, dateTime.minute)
            assertEquals(7, dateTime.second)

            val timestamp = result["created_at_timestamp"] as LocalDateTime

            assertEquals(2020, timestamp.year)
            assertEquals(1, timestamp.monthValue)
            assertEquals(19, timestamp.dayOfMonth)
            assertEquals(3, timestamp.hour)
            assertEquals(14, timestamp.minute)
            assertEquals(7, timestamp.second)

            assertEquals(
                Duration.ofHours(3).plus(Duration.ofMinutes(14).plus(Duration.ofSeconds(7))),
                result["created_at_time"]
            )

            val year = result["created_at_year"] as Short
            assertEquals(1999, year)
        }
    }

    @Test
    fun `it should be able to bind statement values to the prepared statement`() {

        withConnection { connection ->
            val insert =
                """
              insert into numbers (
              number_tinyint,
              number_smallint,
              number_mediumint,
              number_int,
              number_bigint,
              number_decimal,
              number_float,
              number_double
              ) values
              (
              ?,
              ?,
              ?,
              ?,
              ?,
              ?,
              ?,
              ?)
            """

            val byte: Byte = 10
            val short: Short = 679
            val mediumInt = 778
            val int = 875468
            val bigInt = 100007654L
            val bigDecimal = BigDecimal("198.657809")
            val double = 98.765
            val float = 432.8F

            executeQuery(connection, this.createTableNumericColumns)
            executePreparedStatement(
                connection,
                insert,
                listOf(
                    byte,
                    short,
                    mediumInt,
                    int,
                    bigInt,
                    bigDecimal,
                    float,
                    double
                )
            )

            val row = assertNotNull(executePreparedStatement(connection, "SELECT * FROM numbers").rows)[0]

            assertEquals(byte, row["number_tinyint"])
            assertEquals(short, row["number_smallint"])
            assertEquals(mediumInt, row["number_mediumint"])
            assertEquals(int, row["number_int"])
            assertEquals(bigInt, row["number_bigint"])
            assertEquals(bigDecimal, row["number_decimal"])
            assertEquals(float, row["number_float"])
            assertEquals(double, row["number_double"])
        }
    }

    @Test
    fun `bind parameters on a prepared statement`() {

        val create = """CREATE TEMPORARY TABLE posts (
                            id INT NOT NULL AUTO_INCREMENT,
                            some_text TEXT not null,
                            primary key (id) )"""

        val insert = "insert into posts (some_text) values (?)"
        val select = "select * from posts"

        withConnection { connection ->
            executeQuery(connection, create)
            executePreparedStatement(connection, insert, listOf("this is some text here"))
            val row = assertNotNull(executePreparedStatement(connection, select).rows)[0]

            assertEquals(1, row["id"])
            assertEquals("this is some text here", row["some_text"])

            val queryRow = assertNotNull(executeQuery(connection, select).rows)[0]

            assertEquals(1, queryRow["id"])
            assertEquals("this is some text here", queryRow["some_text"])
        }
    }

    @Test
    fun `bind timestamp parameters to a table`() {

        val insert =
            """
          insert into posts (created_at_date, created_at_datetime, created_at_timestamp, created_at_time, created_at_year)
          values ( ?, ?, ?, ?, ? )
        """

        val date = LocalDate.of(2011, 9, 8)
        val dateTime = LocalDateTime.of(2012, 5, 27, 15, 29, 55, 1000)
        val timestamp = Timestamp.valueOf(dateTime)
        val time = Duration.ofHours(3) + Duration.ofMinutes(5) + Duration.ofSeconds(10)
        val year = 2012.toShort()

        withConnection { connection ->
            executeQuery(connection, this.createTableTimeColumns)
            executePreparedStatement(connection, insert, listOf(date, dateTime, timestamp, time, year))
            val rows = assertNotNull(
                executePreparedStatement(
                    connection,
                    "select * from posts where created_at_year > ?",
                    listOf(2011)
                ).rows
            )

            assertEquals(1, rows.size)
            val row = assertNotNull(rows[0])

            assertEquals(date, row["created_at_date"])
            assertEquals(timestamp.toLocalDateTime(), row["created_at_timestamp"])
            assertEquals(time, row["created_at_time"])
            assertEquals(year, row["created_at_year"])
            assertEquals(dateTime, row["created_at_datetime"])
        }
    }

    @Test
    fun `read a timestamp with microseconds`() {

        val create =
            """CREATE TEMPORARY TABLE posts (
       id INT NOT NULL AUTO_INCREMENT,
       created_at_timestamp TIMESTAMP(3) not null,
       created_at_time TIME(3) not null,
       primary key (id)
      )"""

        val insert =
            """INSERT INTO posts ( created_at_timestamp, created_at_time )
           VALUES ( '2013-01-19 03:14:07.019', '03:14:07.019' )"""

        val time = Duration.ofHours(3) +
                Duration.ofMinutes(14) +
                Duration.ofSeconds(7) +
                Duration.ofMillis(19)

        val timestamp = LocalDateTime.of(2013, 1, 19, 3, 14, 7, 19 * 1000_000)
        val select = "SELECT * FROM posts"

        withConnection { connection ->
            executeQuery(connection, create)
            executeQuery(connection, insert)
            val rows = assertNotNull(executePreparedStatement(connection, select).rows)

            val row = assertNotNull(rows[0])

            assertEquals(time, row["created_at_time"])
            assertEquals(timestamp, row["created_at_timestamp"])

            val otherRow = assertNotNull(executeQuery(connection, select).rows)[0]

            assertEquals(time, otherRow["created_at_time"])
            assertEquals(timestamp, otherRow["created_at_timestamp"])
        }
    }

    @Test
    fun `support prepared statement with a big string`() {

        val bigString: String
        val builder = StringBuilder()
        (0..400).map { builder.append("0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789") }

        bigString = builder.toString()

        withConnection { connection ->
            executeQuery(
                connection,
                "CREATE TEMPORARY TABLE BIGSTRING( id INT NOT NULL AUTO_INCREMENT, STRING LONGTEXT, primary key (id))"
            )
            executePreparedStatement(connection, "INSERT INTO BIGSTRING (STRING) VALUES (?)", listOf(bigString))
            val row = assertNotNull(executePreparedStatement(connection, "SELECT STRING, id FROM BIGSTRING").rows)[0]
            assertEquals(1, row["id"])
            val result = row["STRING"] as String
            assertEquals(bigString, result)
        }
    }

    @Test
    fun `support setting null to a column`() {
        val somethingNull: String? = null
        withConnection { connection ->
            executeQuery(
                connection,
                "CREATE TEMPORARY TABLE timestamps ( id INT NOT NULL, moment TIMESTAMP NULL, primary key (id))"
            )
            executePreparedStatement(
                connection,
                "INSERT INTO timestamps (moment, id) VALUES (?, ?)",
                listOf(somethingNull, 10)
            )
            val row = assertNotNull(executePreparedStatement(connection, "SELECT moment, id FROM timestamps").rows)[0]
            assertEquals(10, row["id"])
            assertNull(row["moment"])
        }
    }

    @Test
    fun `support setting None to a column`() {
        val someNull: String? = null
        withConnection { connection ->
            executeQuery(
                connection,
                "CREATE TEMPORARY TABLE timestamps ( id INT NOT NULL, moment TIMESTAMP NULL, primary key (id))"
            )
            executePreparedStatement(
                connection,
                "INSERT INTO timestamps (moment, id) VALUES (?, ?)",
                listOf(someNull, 10)
            )
            val row = assertNotNull(executePreparedStatement(connection, "SELECT moment, id FROM timestamps").rows)[0]
            assertEquals(10, row["id"])
            assertNull(row["moment"])
        }
    }

    @Test
    fun `support setting Some(value) to a column`() {
        withConnection { connection ->
            executeQuery(
                connection,
                "CREATE TEMPORARY TABLE timestamps ( id INT NOT NULL, moment TIMESTAMP NULL, primary key (id))"
            )
            val moment = LocalDateTime.now().withNano(0) // cut off millis to match timestamp
            executePreparedStatement(
                connection,
                "INSERT INTO timestamps (moment, id) VALUES (?, ?)",
                listOf(moment, 10)
            )
            val row = assertNotNull(executePreparedStatement(connection, "SELECT moment, id FROM timestamps").rows)[0]
            assertEquals(10, row["id"])
            assertEquals(moment, row["moment"])
        }
    }

    @Test
    fun `bind parameters on a prepared statement with limit`() {

        val create = """CREATE TEMPORARY TABLE posts (
                            id INT NOT NULL AUTO_INCREMENT,
                            some_text TEXT not null,
                            some_date DATE,
                            primary key (id) )"""

        val insert = "insert into posts (some_text) values (?)"
        val select = "select * from posts limit 100"

        withConnection { connection ->
            executeQuery(connection, create)

            executePreparedStatement(connection, insert, listOf("this is some text here"))

            val row = assertNotNull(executeQuery(connection, select).rows)[0]

            assertEquals(1, row["id"])
            assertEquals("this is some text here", row["some_text"])
            assertNull(row["some_date"])

            val queryRow = assertNotNull(executePreparedStatement(connection, select).rows)[0]

            assertEquals(1, queryRow["id"])
            assertEquals("this is some text here", queryRow["some_text"])
            assertNull(queryRow["some_date"])
        }
    }

    @Test
    fun `insert with prepared statements and without columns`() {
        withConnection { connection ->
            executeQuery(connection, this.createTable)

            executePreparedStatement(connection, this.insert)

            val result = assertNotNull(executePreparedStatement(connection, this.select).rows)
            assertEquals(1, result.size)
            assertEquals("Boogie Man", result[0]["name"])
        }
    }

    @Test
    fun `be able to release prepared statements`() {
        withConnection { connection ->
            val query = "select 1 as id , 'joe' as name"
            val result = executePreparedStatement(connection, query).rows

            assertThat(result[0]("name")).isEqualTo("joe")
            assertThat(result(0)("id")).isEqualTo(1L)
            assertThat(result.size).isEqualTo(1)

            validateCounters(connection, prepare = 1, close = 0)

            val result2 = executePreparedStatement(connection, query).rows

            assertThat(result2[0]("name")).isEqualTo("joe")
            assertThat(result2(0)("id")).isEqualTo(1L)
            assertThat(result2.size).isEqualTo(1)

            validateCounters(connection, prepare = 1, close = 0)

            releasePreparedStatement(connection, query)

            Thread.sleep(2000)

            validateCounters(connection, prepare = 1, close = 1)
        }
    }

    @Test
    fun `be able to release prepared statements immediately`() {
        withConnection { connection ->
            val query = "select 1 as id , 'joe' as name"
            val result = executePreparedStatement(connection, query, emptyList(), true).rows

            assertThat(result[0]("name")).isEqualTo("joe")
            assertThat(result(0)("id")).isEqualTo(1L)
            assertThat(result.size).isEqualTo(1)

            validateCounters(connection, prepare = 1, close = 1)
        }
    }

    private fun validateCounters(connection: MySQLConnection, prepare: Int, close: Int) {
        val statementMetrics = executeQuery(connection, "SHOW SESSION STATUS LIKE 'Com_stmt%'").rows
        assertThat(statementMetrics(3)("Variable_name")).isEqualTo("Com_stmt_prepare")
        assertThat(statementMetrics(3)("Value")).isEqualTo(prepare.toString())
        assertThat(statementMetrics(1)("Variable_name")).isEqualTo("Com_stmt_close")
        assertThat(statementMetrics(1)("Value")).isEqualTo(close.toString())
    }

    @Test
    fun `test interceptor`() {
        try {
            System.setProperty("jasyncDoNotInterceptChecks", "true")
            val interceptor = ForTestingQueryInterceptor()
            MDC.put("a", "b")
            val mdcInterceptor = MdcQueryInterceptorSupplier()
            withConfigurableConnection(ContainerHelper.defaultConfiguration.copy(interceptors = listOf(Supplier<QueryInterceptor> { interceptor }, mdcInterceptor))) { connection ->
                executeQuery(connection, this.createTable)

                awaitFuture(connection.sendPreparedStatement(this.insert).map {
                    assertThat(MDC.get("a")).isEqualTo("b")
                })

                val result = assertNotNull(executePreparedStatement(connection, this.select).rows)
                assertEquals(1, result.size)
                assertEquals("Boogie Man", result[0]["name"])
            }
            assertThat(interceptor.preparedStatements.get()).isEqualTo(2)
            assertThat(interceptor.completedPreparedStatements.get()).isEqualTo(2)
            assertThat(MDC.get("a")).isEqualTo("b")
        } finally {
            System.getProperties().remove("jasyncDoNotInterceptChecks")
        }
    }
}
